pkg/message/lexer.go

```diff
@@ -0,0 +1,161 @@
+package message
+
+import (
+	"errors"
+	"strings"
+)
+
+type Token struct {
+	Value string
+	Level int
+	Type  tokenType
+}
+
+type tokenType int
+
+const (
+	Keyword tokenType = iota
+	PlaceholderOpen
+	PlaceholderClose
+	Literal
+	Text
+	Function
+	Variable
+)
+
+const (
+	Match  = "match"
+	Let    = "let"
+	When   = "when"
+	Dollar = '$'
+	Colon  = ':'
+	Plus   = '+'
+	Minus  = '-'
+)
+
+func Lex(str string) ([]Token, error) {
+	var (
+		tokens           []Token
+		parsedTokens     []Token
+		runes            []rune
+		placeholderLevel int
+	)
+
+	for i, r := range str {
+		switch r {
+		case ' ':
+			if len(runes) > 0 {
+				runes = append(runes, r)
+				tokens = append(tokens, createTokensFromBuffer(runes, placeholderLevel)...)
+
+				runes = []rune{}
+			} else {
+				runes = append(runes, r)
+			}
+		case '{':
+			if len(runes) > 0 {
+				tokens = append(tokens, createTokensFromBuffer(runes, placeholderLevel)...)
+
+				runes = []rune{}
+			}
+
+			placeholderLevel++
+
+			tokens = append(tokens, Token{Type: PlaceholderOpen, Value: "{", Level: placeholderLevel})
+		case '}':
+			if len(runes) > 0 {
+				tokens = append(tokens, createTokensFromBuffer(runes, placeholderLevel)...)
+
+				runes = []rune{}
+			}
+
+			tokens = append(tokens, Token{Type: PlaceholderClose, Value: "}", Level: placeholderLevel})
+			placeholderLevel--
+		case '$', ':', '+', '-':
+			if i+1 < len(str) && str[i+1] == ' ' {
+				return []Token{}, errors.New("variable or function name starts with a space")
+			}
+
+			if len(runes) > 0 {
+				tokens = append(tokens, createTokensFromBuffer(runes, placeholderLevel)...)
+				runes = []rune{}
+			}
+
+			runes = append(runes, r)
+		default:
+			runes = append(runes, r)
+		}
+	}
+
+	if len(runes) > 0 {
+		tokens = append(tokens, createTokensFromBuffer(runes, placeholderLevel)...)
+	}
+
+	parsedTokens, err := combineTextTokens(tokens, parsedTokens)
+	if err != nil {
+		return []Token{}, errors.New("combine Text tokens")
+	}
+
+	return parsedTokens, nil
+}
+
+// Combining Text tokens into one sentence.
+func combineTextTokens(tokens, parsedTokens []Token) ([]Token, error) {
```

Comment:
Function descriptions should start with the function name followed by the description.
example:
// combineTextTokens combines ...